### PR TITLE
upi/vsphere: Use Afterburn guestinfo for static IP and hostname config

### PR DESCRIPTION
**Pull request by LorbusChris**
/test vsphere-upi

### CODE REVIEW

File: upi/vsphere/vm/nm-keyfile.tmpl [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+uuid=${uuid}
+type=802-3-ethernet
+interface-name=ens192
+autoconnect=true
+
+[ipv4]
+method=manual
+dns=${dns}
+dns-search=${cluster_domain}
+addresses=${ip_address}/24
```
Comment (abhinavdahiya) [inline upi/vsphere/vm/nm-keyfile.tmpl:12]: Hmm.. why the fixed network range?

Comment (LorbusChris) [reply upi/vsphere/vm/nm-keyfile.tmpl:12]: it seems to be the default. I haven't seen any examples without it, and we've been using it with this range in OKD as well: https://github.com/openshift/installer/blob/fcos/upi/vsphere/machine/ignition.tf#L41 (that's still TF 0.11x). Happy to change this if needed.

**Comment by LorbusChris** [comment]
It seems the interface isn't coming up. I suspect NM's keyfile plugin is not enabled on RHCOS. Investigating.

**Comment by dustymabe** [comment]
> It seems the interface isn't coming up. I suspect NM's keyfile plugin is not enabled on RHCOS. Investigating.

Try naming the file `"/etc/NetworkManager/system-connections/ens192.nmconnection"`. There are quite a few examples of key files in the test I've been running manually (see https://github.com/coreos/fedora-coreos-config/pull/369)

File: upi/vsphere/vm/nm-keyfile.tmpl (hunk 1 above)
Comment (abhinavdahiya) [reply upi/vsphere/vm/nm-keyfile.tmpl:12]: I'll try to look, but how did we decide the /24, and what does it mean to set it to that value.

Also it seem like we have machine_cidr, how does that relate to this value..

**Comment by LorbusChris** [comment]
@dustymabe @jcpowermac any ideas why this isn't working? Are the boot images maybe not supporting keyfiles, yet?

File: upi/vsphere/vm/ignition.tf [hunk 2]
```diff
@@ -18,12 +18,12 @@ data "ignition_file" "static_ip" {
   for_each = var.hostnames_ip_addresses
 
   filesystem = "root"
-  path       = "/etc/sysconfig/network-scripts/ifcfg-ens192"
+  path       = "/etc/NetworkManager/system-connections/eth0.nmconnection"
```
Comment (jcpowermac) [inline upi/vsphere/vm/ignition.tf:21]: @LorbusChris like @dustymabe mentioned this needs to be `ens192` and in the template.

File: upi/vsphere/vm/nm-keyfile.tmpl [hunk 3]
```diff
@@ -0,0 +1,13 @@
+[connection]
+id=Static ethernet connnection 1
+uuid=${uuid()}
+type=802-3-ethernet
+interface-name=eth0
```
Comment (jcpowermac) [inline upi/vsphere/vm/nm-keyfile.tmpl:5]: `interface-name=ens192`

File: upi/vsphere/vm/ignition.tf (hunk 2 above)
Comment (LorbusChris) [reply upi/vsphere/vm/ignition.tf:21]: I just changed the name in the last commit to test if that changes anything, it was ens192 before and didn't work either :( https://github.com/openshift/installer/pull/3533/commits/1c80ea865b8083e212028c870edf0f8abaae3196

Comment (jcpowermac) [reply upi/vsphere/vm/ignition.tf:21]: Was it like:
"/etc/NetworkManager/system-connections/ens192.nmconnection"
and that didn't work?

I haven't used this method of configuring adapters before so /shrug :)

**Comment by LorbusChris** [comment]
I've rolled back the wip commits. Name is `ens192` again, although that didn't work earlier either.

**Comment by cgwalters** [comment]
Keep in mind this is only going to affect networking in the real root, not at Ignition time.

For that we need the installer to make use of https://github.com/openshift/enhancements/pull/291 specifically https://github.com/coreos/afterburn/pull/404

**Comment by jcpowermac** [comment]
@LorbusChris if this is for master/4.6 maybe we should just use what is documented in #4121 ?

**Comment by LorbusChris** [comment]
I don't think it's a hard requirement for 4.6, but I'm still wondering what the underlying issue here is that prevents the NM keyfile to work.

If what's documented in #4121 is preferable to this I'd be totally fine with closing this PR, though.

File: upi/vsphere/vm/main.tf [hunk 4]
```diff
@@ -29,8 +30,9 @@ resource "vsphere_virtual_machine" "vm" {
   }
 
   extra_config = {
-    "guestinfo.ignition.config.data"          = base64encode(data.ignition_config.ign[each.key].rendered)
-    "guestinfo.ignition.config.data.encoding" = "base64"
+    "guestinfo.afterburn.initrd.network-kargs" = "ip=${each.value}::${cidrhost(machine_cidr, 1)}:255.255.255.0::ens192:off"
```
Comment (jcpowermac) [inline upi/vsphere/vm/main.tf:33]: @LorbusChris the mask isn't correct. Should be: `255.255.254.0`

Comment (LorbusChris) [reply upi/vsphere/vm/main.tf]: @jcpowermac thanks for catching that! updated

**Comment by LorbusChris** [comment]
vsphere UPI failures like flakes to me

/test e2e-vsphere-upi

**Comment by LorbusChris** [comment]
@jcpowermac @staebler any chance you can take a look at this again soon?

**Comment by jcpowermac** [comment]
@LorbusChris upi passed
I think this LGTM

**Comment by staebler** [comment]
My naive reading of [this comment](https://github.com/openshift/installer/pull/3533#issuecomment-647647600) from @cgwalters is that the network settings will not be available at ignition time, which seems to be the same as what we had before with configuring the settings via ignition configs.

/lgtm
/approve

**Comment by rvanderp3** [comment]
Cherry-picking this to 4.7 due to issues found in vSphere UPI CI jobs.  

/cherry-pick release-4.7

**Comment by rvanderp3** [comment]
/retitle Bug 1975527: upi/vsphere: Use Afterburn guestinfo for static IP and hostname config

**Pull request merged**